tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +227,280 @@ static bool HasCanonicalDimensionNumbers(
+
+    // Find input/output values and types.
+    auto loc = op.getLoc();
+    ValueRange inputs = args;
+
+    // Setting up Initial buffer for Output Tensor
+    Value output;
+    auto dyn_sizes = ExtractDynamicEinsumSizes(
+        rewriter, loc, args.front(), args.back(), all_loop_dim[0],
+        all_loop_dim[1], all_loop_dim[2]);
+    output = GetInitTensor(rewriter, loc, *result_ty, dyn_sizes);
+
+    // Create indexing maps.
+    // Create a 1:1 map from f:StrDimension -> affineDimension
+    int64_t nloops = input_ind.size();
+    std::unordered_map<std::string, AffineExpr> str_affine_dim_umap;
+    for (auto it : llvm::enumerate(input_ind)) {
+      str_affine_dim_umap[it.value()] = rewriter.getAffineDimExpr(it.index());
+    }
+    // For each Operands, we generate the affine map for accessing pattern.
+    // Using knowledge from einsum_config/loop of each operand in string type,
+    // we can generate the equivalent information in affine type using
+    // "getExprFromConfig". Then we can feed it into the affinemap builder
+    // where: map = mapBuilder(number_of_dims, num_constant, einsum_config,
+    // mlirCtx)
+    SmallVector<AffineMap, 4> maps;
+    for (auto it : llvm::enumerate(inputs)) {
+      AffineMap scalar_map = AffineMap::get(nloops, 0, rewriter.getContext());
+      auto inputExprs = getExprFromConfig(
+          input_ind, input_loops_dim_vec[it.index()], str_affine_dim_umap);
+      AffineMap multidim_map =
+          AffineMap::get(nloops, 0, inputExprs, rewriter.getContext());
+      maps.push_back(is_scalar(it.value()) ? scalar_map : multidim_map);
+    }
+    auto inputExprs =
+        getExprFromConfig(input_ind, all_loop_dim.back(), str_affine_dim_umap);
+    AffineMap multidim_map =
+        AffineMap::get(nloops, 0, inputExprs, rewriter.getContext());
+    maps.push_back(multidim_map);
+
+    // Creating Regions to be put into linalg.generic
+    // If has reduction axes, operation will become mulf+addf
+    // If no reduction is done, then operation will become just mulf
+    auto reduction_region = [&](OpBuilder& nestedBuilder, Location nestedLoc,
+                                ValueRange args) {
+      Value result_val;
+      auto MulOp =
+          nestedBuilder.create<mlir::MulFOp>(nestedLoc, args[0], args[1]);
+      if (reduction_axe.size() > 0) {
+        auto AddOp = nestedBuilder.create<mlir::AddFOp>(nestedLoc, args[2],
+                                                        MulOp.getResult());
+        result_val = AddOp.getResult();
+      } else
+        result_val = MulOp.getResult();
+      nestedBuilder.create<linalg::YieldOp>(nestedLoc, result_val);
+    };
+
+    // Build `linalg.generic` op.
+    auto linalg_op = rewriter.create<linalg::GenericOp>(
+        loc, result_ty ? *result_ty : TypeRange{}, inputs, output, maps,
+        GetEinsumLoopsAttrs(input_ind, reduction_axe), reduction_region);
+    rewriter.replaceOp(op, linalg_op->getResults());
```

Comment:
[optional] I prefer using `.`.
```suggestion
    rewriter.replaceOp(op, linalg_op.getResults());
```

Reply:
done